Add documentation (#32)
conda-forge/cuda-feedstock: recipe/doc/recipe_guide.md

This guide assumes that you already know how to write and compile code that supports such behavior.
If your library is properly configured as such, you will need to do a bit of extra work to ensure that your conda package supports this as well:

- You must [ignore the `run_exports`](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements) of any CUDA packages that your package depends on. Otherwise, the `run_exports` would require the runtime libraries to have versions equal to or greater than the versions used to build the package.
- You must add explicit runtime dependencies (or `run_constrained` for soft/optional dependencies) that specify the desired version range for the dependencies whose `run_exports` have been ignored. 

As an example, consider that you have built a package that requires `libcublas`:

```yaml
requirements:
  build:

>     - compiler('cuda')

Reviewer: Wrap in `{{` / `}}` since it's a Jinja function.
```suggestion
    - {{ compiler('cuda') }}
```

  host:

>     - libcublas

Reviewer: It's the `-dev` package that has run-exports.
```suggestion
    - libcublas-dev
```

    - cuda-version=12.4
```

By default, at runtime your library will require having the `libcublas` version corresponding to CUDA 12.4 or newer.
To make this compatible with all CUDA 12 minor versions 12.0+, you must add the following:
```yaml
build:
  # Ignore run exports in your build section
  ignore_run_exports_from:
    - {{ compiler('cuda') }}
    - libcublas-dev

requirements:
  run:

>     # Since we’ve ignored the run export, pin manually, but set the min to just "x" since we support any libcublas within the same major release, including older versions

Reviewer: ```suggestion
    # Since we've ignored the run export, we pin manually, but set the min to "x" since we support any libcublas within the same major release, including older versions
```

>     - pin_compatible("libcublas", min_pin="x", max_pin="x")

Reviewer: ```suggestion
    - {{ pin_compatible("libcublas", min_pin="x", max_pin="x") }}
```

```

For packages that need to support both CUDA major versions 11 & 12, you will need to use selectors and/or Jinja tricks to separate out the requirements for CUDA 11 and CUDA 12. [cupy-feedstock](https://github.com/conda-forge/cupy-feedstock) offers a good example.

## Cross-compilation

The CUDA recipes are designed to support cross-compilation.
As such, a number of CUDA components on `conda-forge` are split into `noarch: generic` component packages that are named according to the supported architecture, rather than being architecture-specific packages.
The canonical example is [the cuda-nvcc package](https://github.com/conda-forge/cuda-nvcc-feedstock/blob/main/recipe/meta.yaml) that contains the CUDA `nvcc` compiler.

> This package is split into the `cuda-nvcc` package – which is architecture specific and must be installed on the appropriate target platform (e.g.
> x86-64 Linux) – and the `cuda-nvcc_${TARGET_PLATFORM}` packages – each of which is architecture-independent and may be installed on any target, but are only suitable for use in compiling code for the specified target platform.

Reviewer: ```suggestion
This package is split into the `cuda-nvcc` package -- which is architecture specific and must be installed on the appropriate target platform (e.g.
x86-64 Linux) -- and the `cuda-nvcc_${TARGET_PLATFORM}` packages -- each of which is architecture-independent and may be installed on any target, but are only suitable for use in compiling code for the specified target platform.
```

# Guide for Maintainers of Recipes That Use CUDA

This guide is intended for maintainers of other recipes that depend on CUDA.
It assumes familiarity with the user guides for both [running CUDA code](./end_user_run_guide.md) and [compiling CUDA code](./end_user_compile_guide.md) with conda-forge packages.

## Best Practices

Recipe maintainers are encouraged not to use the metapackages for specifying dependencies, but to instead specify only the minimal subset of CUDA components required for usage of their libraries.

> The `*-dev` variants of the packages all have [`run_exports`](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements) set up such that if your package requires a certain CUDA library at build time, specifying the corresponding dev package as a host requirement will result in your package exporting the corresponding non-dev packages as a runtime dependency.

Author: Yup, typo, thanks.

## CUDA Enhanced Compatibility

Since CUDA 11, [CUDA has offered a number of increased compatibility guarantees](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#cuda-compatibility-developer-s-guide).
The most up to date documentation for these may be found in [the CUDA best practices documentation](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#cuda-toolkit-versioning).

> Of particular interest from a conda packaging perspective is the binary backwards compatibility made by the CTK: packages built with newer versions of the CTK will also run with older minor versions (within the same major family) of the CTK installed, assuming that you have properly guarded any usage of APIs introduced in newer versions with suitable checks.

Reviewer: ```suggestion
Of particular interest from a conda packaging perspective is the binary backwards compatibility made by the CUDA Toolkit (CTK): packages built with newer versions of the CTK will also run with older minor versions (within the same major family) of the CTK installed, assuming that you have properly guarded any usage of APIs introduced in newer versions with suitable checks.
```

This guide assumes that you already know how to write and compile code that supports such behavior.
If your library is properly configured as such, you will need to do a bit of extra work to ensure that your conda package supports this as well:

You must [ignore the run exports](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements) of any CUDA packages that your package depends on.
Otherwise, the `run_exports` would require the runtime libraries to have versions equal to or greater than the versions used to build the package.
You must add explicit runtime dependencies (or `run_constrained` for soft/optional dependencies) that specify the desired version range for the dependencies whose run exports have been ignored. 

As an example, consider that you have built a package that requires libcublas:

```
requirements:
  build:

>     - compiler(‘cuda’)

Reviewer: ```suggestion
    - compiler('cuda')
```

  host:
    - libcublas
    - cuda-version=12.4
```

By default, at runtime your library will require having the libcublas version corresponding to CUDA 12.4 or newer.
To make this compatible with all CUDA 12 minor versions 12.0+, you must add the following:
```
build:
  # Ignore run exports in your build section
  ignore_run_exports_from:
    - {{ compiler('cuda') }}

>     - libcurand-dev

Reviewer: ```suggestion
    - libcublas-dev
```

requirements:
  run:

>     # Since we’ve ignored the run export, pin manually, but set the min to just “x” since we support any libcublas within the same major release, including older versions
>     - pin_compatible(libcublas, min_pin=”x”, max_pin=”x”)

Reviewer: ```suggestion
    # Since we’ve ignored the run export, pin manually, but set the min to just "x" since we support any libcublas within the same major release, including older versions
    - pin_compatible("libcublas", min_pin="x", max_pin="x")
```

```

## Cross-compilation

The CUDA recipes are designed to support cross-compilation.
As such, a number of CUDA components on conda-forge are split into `noarch: generic` component packages that are named according to the supported architecture, rather than being architecture-specific packages.
The canonical example is [the cuda-nvcc package](https://github.com/conda-forge/cuda-nvcc-feedstock/blob/main/recipe/meta.yaml) that contains the CUDA `nvcc` compiler.
This package is split into the `cuda-nvcc` package – which is architecture specific and must be installed on the appropriate target platform (e.g.

> x86-64 Linux) – and the `cuda-nvcc-${TARGET_PLATFORM}` packages – each of which is architecture-independent and may be installed on any target, but are only suitable for use in compiling code for the specified target platform.

Reviewer: It seems like this file is using manual line wraps but the others are not. Let's make sure this is consistent. I prefer semantic line breaks (each sentence on a line) if you need to make a choice.

This approach allows using host machines with a single platform to compile code for multiple platforms.

## Directory structure

### Linux

The conda-forge CUDA packages aim to satisfy two sets of constraints.
On one hand, the packages aim to retain as similar a structure as possible to the CUDA packages that may be installed via system package manager (e.g.
apt).
On the other hand, the packages aim to provide a seamless experience at both build-time and runtime within conda environments.

> To satisfy the first requirement, all files in CUDA conda packages are installed into the `$PREFIX/targets` directory.

Reviewer: Should we clarify that this is really something like `$PREFIX/targets/$TARGET_PLATFORM/`?

This includes libraries, headers, and packaging files, along with other miscellaneous files that may be present in any given package.
To satisfy the second requirement, we symlink a number of these files into standard sysroot locations so that they can be found by standard tooling (e.g.
CMake, compilers, ld, etc).

> Specifically, we apply the following conventions:

Reviewer: Seems like the contents below should be a bulleted list with sub-bullets?

Author: Good catch, looks like Leo took care of it in his comment below.

Shared libraries are symlinked into `$PREFIX/lib`.
This includes the bare name (`libcublas.so`), the SONAME, and the full name.
Pkgconfig files are installed directly into `$PREFIX/lib/pkgconfig`.
These are not symlinked from `$PREFIX/targets`, but are directly moved to this location.
The reason is that pkgconfig files contain relative paths to libraries/headers/etc and the paths cannot be accurate relative to both the `targets` directory and the `lib/pkgconfig` directory.
Since the latter is what `pkgconfig` will use, we choose to install the files into `lib/pkgconfig` and reroot the paths accordingly.
Static libraries and header files are not symlinked into the sysroot directories.
Instead, conda installations of `nvcc` know how to search for these packages in the correct directories.

### Windows

Package structure on Windows.

> Doesn’t have `x64` directory.

Reviewer: Let's just write TODO here if it's going to be so sparse. It's okay to have incomplete docs as long as it's clear that it's incomplete and not just a dead end / lost content.

Author: Agreed, I'll remove.

Reviewer 2: @vyasr By "remove" you meant removing the whole Windows section? It's gone now and it doesn't sound right to me. Windows CUDA packages are fully working.

Author: I haven't removed any indication that Windows is working, I've just removed the section on Windows-specific guidance for recipe maintainers because there wasn't much of note and we can fill that in later.

> Package structure on Windows.
> Doesn’t have `x64` directory.

Reviewer 2: Rewriting:
```suggestion
On Windows, the directory structure is adjusted to align with conda's as well as CUDA on Windows' conventions. Note that the `target` subdirectory does not exist (cross-compilation on Windows is not supported).
```

Reviewer 2: As per https://github.com/conda-forge/cuda-feedstock/pull/32#discussion_r1640182739, this text is gone.

> Library structure, on Windows this would be `%LIBRARY_LIB%` for `.lib` files used during the build `%LIBRARY_BIN%` and `.dll` files used at build time and run time

Reviewer 2: ```suggestion
- Shared libraries (i.e. `.dll`) and executables (i.e. `.exe`) are placed in `%LIBRARY_BIN%`
- Headers are placed in `%LIBRARY_INC%`
- Static libraries and import libraries (i.e. `.lib`) are placed in `%LIBRARY_LIB%`
  - There is an open bug (https://github.com/conda-forge/cuda-cudart-feedstock/issues/28) that might require a change to `%LIBRARY_LIB%\x64`
```

Reviewer 2: As per https://github.com/conda-forge/cuda-feedstock/pull/32#discussion_r1640182739, this text is gone.
